graph/push_v2.go

```diff
@@ -0,0 +1,249 @@
+package graph
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/manifest"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/pkg/progressreader"
+	"github.com/docker/docker/pkg/streamformatter"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/docker/registry"
+	"github.com/docker/docker/runconfig"
+	"github.com/docker/docker/utils"
+)
+
+type v2Pusher struct {
+	*TagStore
+	localRepo Repository
+}
+
+func (p *v2Pusher) Push(localName, repoName string, endpoint registry.APIEndpoint, imagePushConfig *ImagePushConfig, sf *streamformatter.StreamFormatter) (fallback bool, err error) {
+	repo, err := NewV2Repository(repoName, endpoint, imagePushConfig.MetaHeaders, imagePushConfig.AuthConfig)
+	if err != nil {
+		logrus.Debugf("Error getting v2 registry: %v", err)
+		return true, err
+	}
+	return false, p.pushV2Repository(repo, localName, imagePushConfig, sf)
+}
+
+func (p *v2Pusher) getImageTags(localRepo Repository, askedTag string) ([]string, error) {
+	logrus.Debugf("Checking %q against %#v", askedTag, localRepo)
+	if len(askedTag) > 0 {
+		if _, ok := localRepo[askedTag]; !ok || utils.DigestReference(askedTag) {
+			return nil, fmt.Errorf("Tag does not exist for %s", askedTag)
+		}
+		return []string{askedTag}, nil
+	}
+	var tags []string
+	for tag := range localRepo {
+		if !utils.DigestReference(tag) {
+			tags = append(tags, tag)
+		}
+	}
+	return tags, nil
+}
+
+func (p *v2Pusher) pushV2Repository(repo distribution.Repository, localName string, imagePushConfig *ImagePushConfig, sf *streamformatter.StreamFormatter) error {
+	if _, err := p.poolAdd("push", localName); err != nil {
+		return err
+	}
+	defer p.poolRemove("push", localName)
+
+	tags, err := p.getImageTags(p.localRepo, imagePushConfig.Tag)
+	if err != nil {
+		return fmt.Errorf("error getting tags for %s: %s", localName, err)
+	}
+	if len(tags) == 0 {
+		return fmt.Errorf("no tags to push for %s", localName)
+	}
+
+	for _, tag := range tags {
+		if err := p.pushV2Tag(repo, p.localRepo, tag, imagePushConfig.OutStream, sf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *v2Pusher) pushV2Tag(repo distribution.Repository, localRepo Repository, tag string, out io.Writer, sf *streamformatter.StreamFormatter) error {
+	logrus.Debugf("Pushing repository: %s:%s", repo.Name(), tag)
+
+	layerId, exists := localRepo[tag]
+	if !exists {
+		return fmt.Errorf("tag does not exist: %s", tag)
+	}
+
+	// TODO(tiborvass): @dmcgowan isn't this supposed to be outside the tags loop?
+	layersSeen := make(map[string]bool)
+
+	layer, err := p.graph.Get(layerId)
+	if err != nil {
+		return err
+	}
+
+	m := &manifest.Manifest{
+		Versioned: manifest.Versioned{
+			SchemaVersion: 1,
+		},
+		Name:         repo.Name(),
+		Tag:          tag,
+		Architecture: layer.Architecture,
+		FSLayers:     make([]manifest.FSLayer, 0, 4),
```

Comment:
no, should probably just be `[]manifest.FSLayer{}`, I think it was migrated this way